src/cryptsetup/cryptsetup.c

```diff
@@ -810,21 +910,30 @@ static int attach_luks_or_plain_or_bitlk_by_fido2(
         for (;;) {
                 bool processed = false;
 
-                r = acquire_fido2_key(
-                                name,
-                                friendly,
-                                arg_fido2_device,
-                                rp_id,
-                                cid, cid_size,
-                                key_file, arg_keyfile_size, arg_keyfile_offset,
-                                key_data, key_data_size,
-                                until,
-                                arg_headless,
-                                required,
-                                &decrypted_key, &decrypted_key_size,
-                                arg_ask_password_flags);
-                if (r >= 0)
-                        break;
+                if (use_libcryptsetup_plugin == true && !arg_fido2_cid) {
+                        r = attach_luks2_by_fido2(cd, name, until, arg_headless, arg_fido2_device, flags);
+                        if (IN_SET(r, -ENOTUNIQ, -ENXIO, -ENOENT))
+                                return log_debug_errno(SYNTHETIC_ERRNO(EAGAIN),
+                                                       "Automatic FIDO2 metadata discovery was not possible because missing or not unique, falling back to traditional unlocking.");
```

Comment:
hmm, you should leave the loop here, if this worked.

Reply:
I'm a bit confused. What loop do you mean? If you mean waiting for HW token to appear it's still there. I just break from the loop in case -ENOTUNIQ, -ENXIO or -ENOENT happens. -ENOENT means no systemd-fido2 token could be used to unlock LUKS2 device.

Reply:
ah, makes sense now. sorry for the confusion